### ISSUE TITLE
[Feature Request] [Configen] Multiple path patterns

### ISSUE THREAD

**Issue by alexandreMayerowitz** [facebookresearch/hydra#1367]
# 🚀 Feature Request

Hi all!
First of all, thanks for your amazing tool, it simplifies things a lot.
I have a small suggestion to enhance **Configen** tool and makes it more modular. It concerns the **module_path_pattern** schema node.

## Motivation

Currently, to generate some structured configurations based on existing class, we may use **Configen** like this:

```yaml
defaults:
  - configen_schema

configen:
  # output directory
  output_dir: ${hydra:runtime.cwd}

  header: |
    # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
    #
    # Generated by configen, do not edit.
    # See https://github.com/facebookresearch/hydra/tree/master/tools/configen
    # fmt: off
    # isort:skip_file
    # flake8: noqa
  module_path_pattern: 'configen/conf/{{module_path}}.py'

  # list of modules to generate configs for
  modules:
    # Loss functions
    - name: torch.nn.modules.loss
      classes:
        - BCELoss
        - BCEWithLogitsLoss
        - CosineEmbeddingLoss
        - CrossEntropyLoss
        - CTCLoss
        - HingeEmbeddingLoss
        - KLDivLoss
        - L1Loss
        - MarginRankingLoss
        - MSELoss
        - MultiLabelMarginLoss
        - MultiLabelSoftMarginLoss
        - MultiMarginLoss
        - NLLLoss
        - PoissonNLLLoss
        - SmoothL1Loss
        - SoftMarginLoss
        - TripletMarginLoss
        - TripletMarginWithDistanceLoss
```

which creates ``loss.py`` module under ``configen/conf/torch/nn/modules`` folder. Which is fine but it creates, in my own opinion, too many, and unnecessary folders.

For instance, if I also want to add optimizers from Pytorch, like so:

```yaml
defaults:
  - configen_schema

configen:
  # output directory
  output_dir: ${hydra:runtime.cwd}

  header: |
    # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
    #
    # Generated by configen, do not edit.
    # See https://github.com/facebookresearch/hydra/tree/master/tools/configen
    # fmt: off
    # isort:skip_file
    # flake8: noqa
  module_path_pattern: 'configen/conf/{{module_path}}.py'

  # list of modules to generate configs for
  modules:
    # Loss functions
    - name: torch.nn.modules.loss
      classes:
        - BCELoss
        - ...

    # Optimizers
    - name: torch.optim
      classes:
        - Adadelta
        - ...
```

It will results with the following architecture:

```bash
├── configen
    └── conf
         └── torch
               ├── nn
               │   └──  modules
               │          └── loss.py
               └── optim.py

```

## Pitch

**Describe the solution you'd like**

As a solution, I would expect an output like this, which is cleaner:

```bash
├── configen
    └── conf
         ├── losses.py
         └── optimizers.py
```

Which should take the form, in the ``YAML`` file:

```yaml
configen:
  # output directory
  output_dir: ${hydra:runtime.cwd}

  header: ...
  module_path_pattern: 'configen/conf/{{output_module_path}}.py'

  # list of modules to generate configs for
  modules:
    # Loss functions
    - name: torch.nn.modules.loss
      output_module_path: losses
      classes:
        - BCELoss
        - ...

    # Optimizers
    - name: torch.optim
      output_module_path: optimizers
      classes:
        - Adadelta
        - ...
```

**Describe alternatives you've considered**

As an alternative, I can create a ``configen.yaml`` file for each group (i.e. ``configen_losses.yaml``, ``configen_optimizers.yaml``, etc.) and use the appropriate ``module_path_pattern``(i.e. ``configen/conf/losses.py``, etc.).

But in my own opinion, it may be handled in a single configen file. Open to discussion of course. :)

**Comment by jieru-hu** [comment]
Hi @alexandreMayerowitz thank you for the request.

The pitch lgtm. I think we could add this as an extra feature but keep the default as it is now.

Also there is a project ongoing for generating Hydra torch configs here https://github.com/pytorch/hydra-torch using the same tool. Feel free to check it out in case that helps.

cc @omry

**Comment by omry** [comment]
Specifically, there is an effort to standardize torch configs in the repo mentioned by Jieru (cc @romesco).
This is still work in progress and feedback and contributions are welcome.
The idea behind using the same directory for the config structure is that it makes it easier to find the configs. See [importing-convention](https://github.com/pytorch/hydra-torch#importing-convention).

I would rather that we focus on the energy in one place, and the decision to go with this config structure for PyTorch configs was already done.

If you are still interested in perusing this for something else, what do you think about this alternative?

```yaml
  # used if individual modules does not specify their own module_path_pattern
  module_path_pattern: 'gen/{{module_path}}.py 

  # list of modules to generate configs for
  modules:
    - name: torch.nn.modules.loss
      # overrides the global module_path_pattern
      module_path_pattern: 'gen/loses.py
      classes:
        - BCELos
```

**Comment by alexandreMayerowitz** [comment]
Thanks for your answers. Yes I had a look at ``hydra-torch`` and need to have a closer look, with @romesco, to see how to contribute.

The reason of the feature request is to gather thinks. For instance, I can have loss functions that comes from pytorch and some others from an another library (but inheriting from a ``torch.nn.Module``). If configen can help to put all those configs in a single place, I really think it could be of interest.

The solution you proposed @omry sounds good to me, and will perfectly suits the case.

**Comment by omry** [comment]
Gotcha.
Feel free to send a pull to add this to configen (with corresponding tests).

**Comment by romesco** [comment]
@mayeroa If you are still interested in contributing to `hydra-torch`, please let me know! Your help will definitely speed up the process! If you want to talk with higher frequency, hop on https://hydra-framework.zulipchat.com and post in the #Hydra Torch channel. I can help you get up and running with contributing!

**Comment by omry** [comment]
Thanks for the proposal and the discussion. The idea remains reasonable, but Configen has not seen meaningful feature development in several years, and there has not been enough community interest to justify expanding it further.

I am closing this as not planned. If interest returns, a new focused proposal or pull request is welcome.
